src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -469,6 +583,79 @@ private static IEnumerable<string> GetPossibleMatchesForSelectedOperation(IEdmOp
             yield return qualifiedContainerName + operationNameWithParameters;
         }
 
+        private bool IsValidExpandToken(string item)
+        {
+            // Check #1: Expand token must contain '(' and end with ')', must be related to navigation property.
+            int idxLP = item.IndexOf('(');
+            if (idxLP == -1
+                || !item.EndsWith(")", StringComparison.Ordinal)
+                || (this.structuredType != null && !IsNavigationPropertyToken(item.Substring(0, idxLP))))
+            {
+                // selected item is not properly parenthesized, or is an operation token.
+                return false;
+            }
+
+            // Check #2: parentheses must be balanced.
+            int count = 0;
+            foreach (char c in item)
+            {
+                if (c == '(')
+                {
+                    ++count;
+                }
+                else if (c == ')')
+                {
+                    --count;
+                }
+
+                if (count < 0)
+                {
+                    // Malformed parentheses detected ==> not expand token.
+                    return false;
+                }
+            }
+
+            return count == 0;
+        }
+
+        /// <summary>
+        /// Checks the specified model and structuredType and see whether the token can be resolved to a navigation property name.
+        /// </summary>
+        /// <param name="token">The token to check.</param>
+        /// <returns>true if token can be resolved to a navigation property; false otherwise.</returns>
+        private bool IsNavigationPropertyToken(string token)
+        {
+            const char nameSpaceSeparator = '.';
+
+            /* Decision tree:
+             #1. if it matches a defined navigation property => treat it as a navigation property
+             #2. otherwise, if the name contains a dot => it's not a navigation property
+             #3. otherwise, if it matches an unqualified bound operation name => it's not a navigation property
+             #4. otherwise, it's a navigation property
+             */
+
+            // For better readability, set the value in if-else branches corresponding to decision tree above.
+            bool found;
+            if (this.structuredType.NavigationProperties().Any(_ => _.Name.Equals(token, StringComparison.Ordinal)))
+            {
+                // #1
+                found = true;
+            }
+            else if (token.IndexOf(nameSpaceSeparator) != -1 ||
+                     this.edmModel.FindBoundOperations(this.structuredType).Any(op => op.Name.Equals(token, StringComparison.Ordinal)))

```

Comment:
>op.Name.Equals(token, StringComparison.Ordinal) [](start = 86, length = 47)

This should also use case sensitivity of UriResolver.

Reply:
Actually, as Biao points out, this in the contextUrl which is case-normalized.

---
In reply to: [227065902](https://github.com/OData/odata.net/pull/1286#discussion_r227065902) [](ancestors = 227065902)